osu.Game/Screens/Play/GameplayClockContainer.cs

```diff
@@ -225,8 +222,8 @@ private void updateRate()
             speedAdjustmentsApplied = true;
             track.ResetSpeedAdjustments();
 
-            track.AddAdjustment(AdjustableProperty.Frequency, pauseFreqAdjust);
-            track.AddAdjustment(AdjustableProperty.Tempo, UserPlaybackRate);
+            (track as IAdjustableAudioComponent)?.AddAdjustment(AdjustableProperty.Frequency, pauseFreqAdjust);
+            (track as IAdjustableAudioComponent)?.AddAdjustment(AdjustableProperty.Tempo, UserPlaybackRate);
 
             foreach (var mod in mods.OfType<IApplicableToTrack>())
                 mod.ApplyToTrack(track);
```

Comment:
I'm not quite sure about this one but I think it's worth mentioning: since `AllowRateAdjustments` is forced to `true` on `Player`, and now that `MusicController` is moved to an `OsuGameBase` level, this logic can be removed, right?

Reply:
I'm hesitant to make changes in this PR. This doesn't immediately work because of the `Reset()` call(s). The following demonstrates a way to make this work, but there's quite a few things to consider (e.g what was the deadlock we were experiencing?):

```diff
diff --git a/osu.Game/Screens/Play/GameplayClockContainer.cs b/osu.Game/Screens/Play/GameplayClockContainer.cs
index f3466a562..cb3eb429e 100644
--- a/osu.Game/Screens/Play/GameplayClockContainer.cs
+++ b/osu.Game/Screens/Play/GameplayClockContainer.cs
@@ -122,13 +122,10 @@ private void load(OsuConfigManager config)

         public void Restart()
         {
-            // The Reset() call below causes speed adjustments to be reset in an async context, leading to deadlocks.
-            // The deadlock can be prevented by resetting the track synchronously before entering the async context.
-            track.ResetSpeedAdjustments();
-
             Task.Run(() =>
             {
-                track.Reset();
+                track.Seek(0);
+                track.Stop();

                 Schedule(() =>
                 {
@@ -216,14 +213,13 @@ protected override void Update()

         private void updateRate()
         {
-            speedAdjustmentsApplied = true;
-            track.ResetSpeedAdjustments();
+            if (speedAdjustmentsApplied)
+                return;

             (track as IAdjustableAudioComponent)?.AddAdjustment(AdjustableProperty.Frequency, pauseFreqAdjust);
             (track as IAdjustableAudioComponent)?.AddAdjustment(AdjustableProperty.Tempo, UserPlaybackRate);

-            foreach (var mod in mods.OfType<IApplicableToTrack>())
-                mod.ApplyToTrack(track);
+            speedAdjustmentsApplied = true;
         }

         protected override void Dispose(bool isDisposing)
@@ -234,11 +230,13 @@ protected override void Dispose(bool isDisposing)

         private void removeSourceClockAdjustments()
         {
-            if (speedAdjustmentsApplied)
-            {
-                track.ResetSpeedAdjustments();
-                speedAdjustmentsApplied = false;
-            }
+            if (!speedAdjustmentsApplied)
+                return;
+
+            (track as IAdjustableAudioComponent)?.RemoveAdjustment(AdjustableProperty.Frequency, pauseFreqAdjust);
+            (track as IAdjustableAudioComponent)?.RemoveAdjustment(AdjustableProperty.Tempo, UserPlaybackRate);
+
+            speedAdjustmentsApplied = false;
         }
```